File: src/main/java/tasks/RuCode/rucode2023/taskL.py
import heapq

# inputFile = open('001.in', 'right')
# outputFile = open('001lol.out', 'w')
#
#
# def input():
#     return inputFile.readline()
#
#
# def print(s):
#     outputFile.write(s)
#     outputFile.write('\n')
#
#
# def printL():
#     outputFile.write('\n')


printL = print

N = int(input())
p = list(map(int, input().split()))
T = int(input())

printL()
printL()
printL()

neg_heap_saved = [(-v, i) for i, v in enumerate(p) if v > 0]
heapq.heapify(neg_heap_saved)
for i in range(T):
    if i > 0:
        input()
    neg_heap = neg_heap_saved.copy()
    retest_result = '0'
    while neg_heap and retest_result == '0':
        _, ind = heapq.heappop(neg_heap)

        print(f'retest {ind + 1}')
        printL()
        input()
        retest_result = input().strip()

    print('done')

File: src/main/java/tasks/RuCode/rucode2023/TaskL.java
package tasks.RuCode.rucode2023;

import java.util.*;
import java.io.*;

public class TaskL {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(reader.readLine().trim());
        int[] p = Arrays.stream(reader.readLine().trim().split(" ")).mapToInt(Integer::parseInt).toArray();
        int T = Integer.parseInt(reader.readLine().trim());

        List<int[]> negHeapSavedList = new ArrayList<>();
        for (int i = 0; i < p.length; i++) {
            if (p[i] > 0) {
                negHeapSavedList.add(new int[]{-p[i], i});
            }
        }
        PriorityQueue<int[]> negHeapSaved = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        negHeapSaved.addAll(negHeapSavedList);

        for (int t = 0; t < T; t++) {
            PriorityQueue<int[]> negHeap = new PriorityQueue<>(negHeapSaved);
            String retestResult = "0";
            while (!negHeap.isEmpty() && retestResult.equals("0")) {
                int ind = negHeap.poll()[1];

                System.out.println("retest " + (ind + 1));
                System.out.flush();
                reader.readLine();
                retestResult = reader.readLine().trim();
            }
            System.out.println("done");
            System.out.flush();
            reader.readLine();
        }
    }
}
